ada/language/ast.py

```diff
@@ -467,32 +467,48 @@ def sub_equation():
                   Entity.stop_resolution_equation,
                   Entity.xref_equation)
 
-    @langkit_property(return_type=Bool, dynamic_vars=[env, origin])
-    def resolve_names_internal(initial=Bool, additional_equation=Equation):
+    @langkit_property(return_type=Bool, memoized=True, call_memoizable=True,
+                      dynamic_vars=[env, origin])
+    def resolve_own_names():
         """
-        Internal helper for resolve_names, implementing the recursive logic.
+        Internal helper for resolve_names. Resolve names for this node up to
+        xref_entry_point and xref_stop_resolution boundaries.
         """
+        return Entity.xref_equation.solve
 
-        solve_xref = Var(initial | Self.xref_stop_resolution)
-
-        i = Var(If(
-            solve_xref,
-            (Entity.xref_equation & additional_equation).solve,
-            True)
-        )
-
-        j = Var(Self.children.all(lambda c: c.then(
-            # Do not explore nodes that are xref entry points, and are not the
-            # initial node.
+    @langkit_property(return_type=Bool, dynamic_vars=[env, origin])
+    def resolve_children_names():
+        """
+        Internal helper for resolve_names, implementing the recursive logic
+        needed to resolve names across xref_stop_resolution boundaries.
+        """
+        return Self.children.all(lambda c: c.then(
+            # Only resolve nodes that have xref_stop_resolution set, and do not
+            # recursively explore nodes that are xref entry points.
             lambda c: If(
                 c.xref_entry_point,
                 True,
-                c.as_entity.resolve_names_internal(False, LogicTrue()),
+                If(c.xref_stop_resolution, c.as_entity.resolve_own_names, True)
+                & c.as_entity.resolve_children_names,
             ),
             default_val=True
-        )))
+        ))
 
-        return i & j
+    @langkit_property(return_type=Bool, dynamic_vars=[env, origin])
+    def resolve_names_internal():
+        """
+        Resolves names for this node up to xref_entry_point boundaries.
+        """
+        return Entity.resolve_own_names & Entity.resolve_children_names
+
+    @langkit_property(return_type=Bool, dynamic_vars=[env, origin])
+    def resolve_names_internal_with_eq(additional_equation=Equation):
```

Comment:
I'm curious: is there a reason you decided to split the implem in two ?

Reply:
Yes: `resolve_names_internal` can take advantage of memoization because it calls resolve_own_name to resolve self's names. `resolve_names_internal_with_eq` cannot be memoized because the additional equation is not hashable, so here the work will always have to be done in full, even when called with `LogicTrue()`.